### PR TITLE
Add GpxAnimator to replay .gpx traces

### PR DESCRIPTION
**Pull request by danesfeder**
This adds a `GpxAnimator` class that subclasses `TimeAnimator`.  The class is fed an `InputStream` and callback that's fired once the input stream is done being parsed.  At that point, `GpxAnimator#start()` can be called and `Location` updates will begin to be provided. 

### Example: 
```
    try {
      InputStream in = getAssets().open("morning_row.gpx");
      GpxAnimator gpxAnimator = new GpxAnimator(in, TimeAnimator::start);
      gpxAnimator.addLocationListener(gpxLocation -> Timber.d("GpxLocation received: %s", gpxLocation));
     } catch (IOException e) {
      e.printStackTrace();
    }
```

### CODE REVIEW

File: liblocation/src/main/java/com/mapbox/android/core/location/GpxAnimator.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * with {@link GpxAnimator#start()}.
+ *
+ * The {@link InputStream} will be automatically closed once the parsing has finished.
+ *
+ * @since 0.3.0
+ */
+@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
+public class GpxAnimator extends TimeAnimator {
+
+  private static final String TAG = GpxAnimator.class.getSimpleName();
```
Comment (tobrun) [inline liblocation/src/main/java/com/mapbox/android/core/location/GpxAnimator.java]: users in the past have reached out about the `class.getSimpleName()` construct being expensive as it uses reflection

File: liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    Document document = documentBuilder.parse(inputStream);
+    Element elementRoot = document.getDocumentElement();
+
+    NodeList trackPointNodes = elementRoot.getElementsByTagName(TAG_TRACK_POINT);
+    if (trackPointNodes == null || trackPointNodes.getLength() == 0) {
+      return null; // Gpx trace did not contain correct tagging
+    }
+
+    try {
+      return createGpxLocationList(trackPointNodes);
```
Comment (tobrun) [inline liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java]: can we use try with resources instead? can the IOException be handled by the end-developer themselves?

File: liblocation/src/main/java/com/mapbox/android/core/location/OnGpxAnimatorReadyCallback.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.mapbox.android.core.location;
+
+/**
+ * Callback passed into the contructor of {@link GpxAnimator}.
+ * <p>
+ * This callback will be triggered once the GPX parsing of the {@link java.io.InputStream} has finished.
+ * <p>
+ * At this point, it is safe to call {@link GpxAnimator#start()}, beginning location updates.
+ *
```
Comment (tobrun) [inline liblocation/src/main/java/com/mapbox/android/core/location/OnGpxAnimatorReadyCallback.java]: missing two times a `</p>`

File: liblocation/src/test/resources/test.gpx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                <ele>4.94</ele>
+                <time>2009-10-17T18:37:31Z</time>
+            </trkpt>
+            <trkpt lat="47.644002" lon="-122.326002">
+                <ele>6.87</ele>
+                <time>2009-10-17T18:37:34Z</time>
+            </trkpt>
+        </trkseg>
+    </trk>
+</gpx>
```
Comment (tobrun) [inline liblocation/src/test/resources/test.gpx]: have we tested that it parses gpx files correctly other as above?

**Comment by electrostat** [comment]
Looks good to me except for the comments by @tobrun ☝️ 

There is a known issues with `roboelectric-3.8` that overwrites some methods in `com.android.support.test`. Bumped to `4.0-alpha-2` and that resolved the problem.

File: liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    } finally {
+      inputStream.close();
+    }
+  }
+
+  @NonNull
+  private List<Location> createGpxLocationList(NodeList trackPointNodes) throws ParseException {
+    List<Location> gpxLocations = new ArrayList<>();
+
+    for (int i = 0; i < trackPointNodes.getLength(); i++) {
```
Comment (andrlee) [inline liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java]: should we simplify this loop down to for-each?

File: liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @NonNull
+  private Double createCoordinate(NamedNodeMap attributes, String attributeName) {
+    String coordinateTextContent = attributes.getNamedItem(attributeName).getTextContent();
+    return Double.parseDouble(coordinateTextContent);
+  }
+
+  @NonNull
+  private Long createTime(Node trackPoint) throws ParseException {
```
Comment (andrlee) [inline liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java]: Should we defer to primitives as opposed to using boxed types?

File: liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      if (node.getNodeName().contains(TAG_TIME)) {
+        Date date = dateFormat.parse(node.getTextContent());
+        return date.getTime();
+      }
+    }
+    return 0L;
+  }
+
+  @NonNull
+  private Location buildGpxLocation(Double latitude, Double longitude, Long time) {
```
Comment (andrlee) [inline liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java]: Should we defer to primitives as opposed to using boxed types?

File: liblocation/src/test/java/com/mapbox/android/core/location/GpxTimeListenerTest.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static junit.framework.Assert.assertNotNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(RobolectricTestRunner.class)
```
Comment (andrlee) [inline liblocation/src/test/java/com/mapbox/android/core/location/GpxTimeListenerTest.java]: what are the benefits of using RobolectricTestRunner vs. MockitoJUnitRunner?

**Comment by alfwatt** [comment]
Stale

**Pull request closed without merging**